File: Leetcode/Python3/179. Largest Number/solution.py
# https://leetcode.com/problems/largest-number/


from typing import List


class Solution:  # Insertion sort
    def check_swap(self, num1: int, num2: int) -> bool:
        return str(num1) + str(num2) < str(num2) + str(num1)

    def largestNumber(self, nums: List[int]) -> str:
        idx = 1  # 0 ~ idx - 1 is sorted
        while idx < len(nums):
            temp = idx
            """
            ex. nums[temp - 1] = 9, nums[temp] = 10
            Check 910 < 109 
            If false not move numbers than continue
            """
            while temp > 0 and self.check_swap(nums[temp - 1], nums[temp]):
                nums[temp], nums[temp - 1] = nums[temp - 1], nums[temp]
                temp -= 1
            idx += 1
        # Convert to int to eliminate cases start with zero
        return str(int("".join(map(str, nums))))


class Solution:  # Optimize where delete zero
    def largestNumber(self, nums: List[int]) -> str:
        nums = list(map(str, nums))
        sort_index = 1
        while (sort_index < len(nums)):
            index = sort_index
            while index > 0 and self.compare(nums[index - 1], nums[index]):
                nums[index], nums[index - 1] = nums[index - 1], nums[index]
                index -= 1
            sort_index += 1
        result = "".join(nums)

        return result if result[0] != "0" else "0"

    def compare(self, num1: str, num2: str) -> bool:
        return num1 + num2 < num2 + num1

File: Leetcode/C++/179. Largest Number/solution.cpp
// https://leetcode.com/problems/largest-number/description/

#include <algorithm>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

class Solution // Optimized code
{
public:
    string largestNumber(vector<int> &nums)
    {
        vector<string> strNums;
        for (auto num : nums)
        {
            strNums.push_back(to_string(num));
        }
        // Using a lambda function for sorting
        sort(strNums.begin(), strNums.end(), [](string num1, string num2)
             { return num1 + num2 > num2 + num1; });

        string result = "";
        for (auto s : strNums)
        {
            result += s;
        }
        return result[0] == '0' ? "0" : result;
    }
};

class Solution // First code
{
public:
    string largestNumber(vector<int> &nums)
    {
        int sort_index = 1;
        while (sort_index < nums.size())
        {
            int index = sort_index;
            /*
            ex. nums[temp - 1] = 9, nums[temp] = 10
            Check 910 < 109
            If false not move numbers than continue
            */
            while (index > 0 && compare(to_string(nums[index - 1]), to_string(nums[index])))
            {
                swap(nums[index - 1], nums[index]);
                index--;
            }
            sort_index++;
        }
        // Move vector data to string
        stringstream ss;
        for (auto num : nums)
        {
            ss << to_string(num);
        }
        string result = ss.str();
        return result[0] == '0' ? "0" : result;
    }

    bool compare(string num1, string num2)
    {
        return num1 + num2 < num2 + num1;
    }
};
